onnxruntime/core/session/provider_registration.cc

```diff
@@ -108,11 +146,45 @@ ORT_API_STATUS_IMPL(OrtApis::SessionOptionsAppendExecutionProvider,
                                  (std::string(provider_name) + " execution provider is not supported in this build. ").c_str());
   };
 
+  auto create_unknown_provider_status = [&provider_name](gsl::span<const EpToAppend> supported_eps) -> OrtStatus* {
+    std::ostringstream str_builder;
+    str_builder << "Unknown provider name '" << provider_name << "'. "
+                << "Currently supported values are ";
+    const size_t num_eps = supported_eps.size();
+    for (size_t i = 0; i < num_eps; ++i) {
+      const EpToAppend& ep_info = supported_eps[i];
+
+      str_builder << "'" << ep_info.short_name << "'/'" << ep_info.canonical_name << "'";
+      if (num_eps >= 2 && i == num_eps - 2) {
+        str_builder << ", and ";
+      } else if (i == num_eps - 1) {
+        str_builder << ".";
+      } else {
+        str_builder << ", ";
+      }
+    }
+
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, str_builder.str().c_str());
+  };
+
+  auto ep_to_append_iter = std::find_if(supported_eps.begin(), supported_eps.end(),
+                                        [&provider_name](const EpToAppend& elem) -> bool {
+                                          return (strcmp(provider_name, elem.short_name) == 0) ||
+                                                 (strcmp(provider_name, elem.canonical_name) == 0);
+                                        });
+
+  if (ep_to_append_iter == supported_eps.end()) {
+    return create_unknown_provider_status(supported_eps);
+  }
+
+  const EpToAppend& ep_to_append = *ep_to_append_iter;
+  ORT_ENFORCE(ep_to_append.id != EpID::INVALID);
+
   // Add provider options to the session config options.
   // Use a new key with the format: "ep.<EP_NAME>.<PROVIDER_OPTION_KEY>"
   for (const auto& [key, value] : provider_options) {
     std::ostringstream new_key_builder;
-    new_key_builder << "ep." << provider_name << "." << key;
+    new_key_builder << "ep." << ep_to_append.canonical_name << "." << key;
```

Comment:
Could we make the prefix all lowercase so it's not case sensitive? 

Longer term these values will come from plugin EPs and a case-insensitive treatment might be slightly more user-friendly.

Reply:
Yeah, I made this lowercase now.